### ISSUE TITLE
Factory Reset Protection

### ISSUE THREAD

**Issue by Stan805** [GrapheneOS/os-issue-tracker#4557]
Implement some kind of protection against unauthorized factory resets through recovery, like if its reset via recovery, making it so that the OOBE asks for the phone password that was set before the reset. This could prevent people from just using or selling a stolen phone.

**Comment by MinePossu** [comment]
One way i could see having a sort of frp is a timer. Like after 7 days it could be fully up and running in case forgetting a password.

**Comment by 2004seraph** [comment]
It's admittedly pretty far out, but a way I can think of more safely implementing this is by having a similar thing to Google's phone number verification which automatically ties your number to your google account.

FRP could be an opt-in mechanism wherein your device periodically pings the GrapheneOS server with your current phone number, constantly keeping a key-value store of `hardware identifiers => phone numbers` up to date, say, every six hours. The FRP partition would have a flag stating FRP is in use, and some data associated with a recovery code.

During OOBE, GrapheneOS could check if the FRP partition has the flag stating its in use, require the device to be online for this portion of the setup if so, and then if the device is actually registered in this database, require either an SMS code or an automatic configuration message (if that SIM is already installed) to be supplied in order to continue. This would essentially tie FRP to your SIM PIN (you'd want to enforce setting this up in order to use FRP) assuming your phone is stolen with it installed, which makes sense since SIM data persists between factory resets.

There are a number cons off the top of my head:
- It requires cellular connectivity:
  - The GrapheneOS project has to pay for a text gateway service.
  - The users have to sacrifice privacy, and also pay for cellular connectivity (a phone number that can receive SMSs at the very least). This feature would be unusable for a portion of GrapeneOS users.
- After a long time not using the phone, you may no longer own the number. Perhaps this could be mitigated with a 2FA-style recovery code that overrides the phone number verification during OOBE, it would make sense for this recovery code to also bypass the internet connectivity requirement. 
- This is some incredibly sensitive data that the GrapheneOS server-side would need to hold onto, but it would be opt-in. Personally, I feel I already trust the maintainers enough to do this.

Why I'm not too fussed about the cons: I think the cellular requirement is a necessary evil, by opting in, you potentially reduce your security and privacy (and tie your FRP to your bills), but in the end I think this is far less likely to brick phones than a passcode you enter when you buy your phone and forget a year later. As you say, FRP is not a security measure. So even though this solution would be possible to bypass by someone dedicated enough, I believe it adds a sufficient amount of difficulty in extracting value from the theft, which is something I would be willing to trade some privacy for. Especially with a strong SIM PIN. 

An even more fanciful idea is that maybe you *could* facilitate FRP recovery (in the event both the recovery code and phone number are lost) by allowing those who opted in to show a phone bill proving they owned that number at the time the device was pinging the server. I've ignored the obvious privacy-related issues and difficulty in being consistent across regions. A mitigation for this could just be encouraging those who opt-in, to pay a carrier anonymously, if they do not trust the GrapheneOS maintainers.

I'd love to hear some other ideas about this. I thought this up in the morning, so there's probably many angles and edge cases I've missed. I had a lot of fun thinking about this

**Comment by imthosea** [comment]
> We'd also likely need to make our own account system with at least email recovery...

What if it could let you sign in with oauth so you can verify with your Google account? That'd take care of recovery methods.

**Comment by JonyBepary** [comment]
I think the current suggestions are leading to a mediocre solution. This needs to be handled with a strictly cryptographic approach.

The most logical solution is a custom offline Anti-Theft Protocol. Unlike Google’s FRP, which relies on central servers for authentication, this method will use a **physical cryptographic token** (stored on Paper or USB) held solely by the user.

I have a good idea, I'll do some digging, share with you guys later.

**Comment by Stan805** [comment]
I wouldn't mind the option of a self-hosted server doing the FRP enforcement.

**Comment by climbinginceptionbee** [comment]
> The ability to use a passkey for FRP

I think this is really the only viable solution, specifically with a physical FIDO2 security key, this should also require two keys to avoid the scenario of someone losing one of them bricking their device. 

In saying this though, I don't see this as a particularly _useful_ feature for most people as it currently stands given that physical security keys aren't particularly widely adopted by the general population

>  **physical cryptographic token**

I could potentially see this as feasible given that its already how backups are done via seedvault

**Comment by schklom** [comment]
The situation changed slightly since the last comment.
> Google's approach works because Google accounts have recovery via a recovery phone, recovery email or customer support if neither of those is available as an option

Since there is now a partnership with Motorola, could they be interested to provide FRP via their [account system](https://account.motorola.com/us/en/account/login/index.html)? They [advertise](https://motorolanews.com/motorola-three-new-b2b-solutions-at-mwc-2026/) "enterprise portfolio", and having FRP sounds pretty useful for corporate environments.

Although they don't have a recovery via customer support or phone number, they allow logging in with Google, Microsoft, and Facebook accounts, so a user could benefit from Google's recovery methods to access their Motorola account.
The main thing is GrapheneOS wouldn't be responsible for FRP accounts.

If implemented, I imagine this would be opt-in during the Setup Wizard and an interface in the System Settings to enable it if a Motorola account is present?

A more general implementation could be to allow for any account listed on `Settings` -> `Passwords, passkeys and accounts` to provide FRP, maybe add a warning, and the user would be responsible for choosing one with a good recovery process. That way, a Google/Microsoft/Motorola/Signal/... account could be used for FRP.

**Comment by thestinger** [comment]
We aren't going to do that. It's going to be the same GrapheneOS regardless of device and we're only going to have our own services.

**Comment by Scriptifer** [comment]
I'm not sure if this is possible, but if the bootloader is locked, and there is no FRP, then can't someone factory reset the device, to then enable OEM unlocking after, unlock the bootloader and flash a modified GrapheneOS with their own custom keys to sign and relock? An evil maid attack would be then possible and log all attempts? Correct me if I'm wrong.

**Comment by Scriptifer** [comment]
> @Scriptifer They could also switch the device for another, so what's your point?

Yes I'm aware, so can't you physically draw something small or leave a mark at the back of the device so that if you get it back, you can check if the mark is there, and if not, then it was switched, or check the serial number.

**Comment by thestinger** [comment]
@Scriptifer They would be permanently preventing recovery of any of your data by doing it so they wouldn't be able to use your password to decrypt anything. It's a good thing if someone wipes your device. If they wanted your data then they could give you another device of the same model with a malicious OS and hope you don't notice the verified boot key fingerprint is different and then use the captured password in the original device. They could also have tampered with the hardware of either a different device or your device. They could have set up intercepting input/output for the touchscreen. You're describing a way of doing it where they can't get the user's data since they wiped the secrets needed to recover it. Auditor could protect against the device being swapped to an extent if it had a quick tile or lockscreen shortcut mode unless they're sophisticated enough to set up automatically proxying it.

**Comment by thestinger** [comment]
@Scriptifer What's the goal of wiping a device and giving it back? Even if they set it up to capture the password with a malicious OS, the data will be unrecoverable with your PIN/password even if they copied the data off the SSD beforehand. You could detect the malicious OS via the verified boot key fingerprint if you were checking. That's assuming they aren't sophisticated enough to tamper with the hardware to intercept input/output for the touchscreen. Auditor could confirm both the device is genuine and the pairing matches before unlocking if it had a quick tile or probably better a lockscreen shortcut for this, but a sophisticated attacker could proxy it to the original device.

The attack you're describing by wiping the device and returning it lacks a clear purpose and they'd have to be too unsophisticated to mess with the touchscreen, etc. but sophisticated enough to make a malicious OS to steal your password which they couldn't use recover the data since they wiped it. An attacker wiping your data for you is generally desirable and it's highly unusual for them to want to wipe your data for you if they're targeting you.
